File: packages/model-providers/src/factory.ts
import Anthropic from "@anthropic-ai/sdk";
import OpenAI from "openai";
import type { ModelProvider } from "@opentalos/core-types";
import { createAnthropicProvider, type AnthropicClientLike } from "./anthropic.js";
import { createOpenAICompatibleProvider, type OpenAIClientLike } from "./openai-compatible.js";
import { createOllamaProvider } from "./ollama.js";
import { createMockProvider } from "./mock.js";

const CHINESE_PROVIDER_PRESET_BASE_URLS: Record<string, string> = {
  dashscope: "https://dashscope.aliyuncs.com/compatible-mode/v1",
  doubao: "https://ark.cn-beijing.volces.com/api/v3",
  kimi: "https://api.moonshot.cn/v1",
  minimax: "https://api.minimax.chat/v1",
};

const KNOWN_PROVIDERS = ["anthropic", "openai-compatible", "ollama", "mock", ...Object.keys(CHINESE_PROVIDER_PRESET_BASE_URLS)];

function requireEnv(name: string): string {
  const value = process.env[name];
  if (!value) {
    throw new Error(`${name} environment variable is required for the selected MODEL_PROVIDER`);
  }
  return value;
}

/** Builds the ModelProvider this deployment uses, from environment variables. Fails fast (throws)
 * on missing/invalid configuration rather than silently falling back — mirrors the ADMIN_API_KEY
 * pattern in apps/api: running with a misconfigured or absent model provider is a real deployment
 * error, not something to paper over with a default. */
export function createModelProviderFromEnv(): ModelProvider {
  const providerName = requireEnv("MODEL_PROVIDER");
  if (!KNOWN_PROVIDERS.includes(providerName)) {
    throw new Error(`Unknown MODEL_PROVIDER "${providerName}" — expected one of: ${KNOWN_PROVIDERS.join(", ")}`);
  }

  if (providerName === "mock") {
    return createMockProvider();
  }

  const model = requireEnv("MODEL_NAME");

  if (providerName === "anthropic") {
    const apiKey = requireEnv("MODEL_API_KEY");
    const client = new Anthropic({ apiKey, baseURL: process.env.MODEL_BASE_URL }) as unknown as AnthropicClientLike;
    return createAnthropicProvider(client, { model });
  }

  if (providerName === "ollama") {
    const baseUrl = requireEnv("MODEL_BASE_URL");
    return createOllamaProvider(fetch, { baseUrl, model });
  }

  // "openai-compatible" and every Chinese-provider preset all route through the same real
  // OpenAI-SDK-backed client, differing only in which base URL they default to.
  const apiKey = requireEnv("MODEL_API_KEY");
  const baseURL = process.env.MODEL_BASE_URL ?? CHINESE_PROVIDER_PRESET_BASE_URLS[providerName];
  const client = new OpenAI({ apiKey, baseURL }) as unknown as OpenAIClientLike;
  return createOpenAICompatibleProvider(client, { model });
}

File: services/memory-service/providers/factory.py
import os

import httpx
from anthropic import AsyncAnthropic
from openai import AsyncOpenAI

from providers.anthropic_provider import AnthropicProvider
from providers.base import TextCompletionProvider
from providers.mock_provider import MockProvider
from providers.ollama_provider import OllamaProvider
from providers.openai_compatible_provider import OpenAICompatibleProvider

CHINESE_PROVIDER_PRESET_BASE_URLS = {
    "dashscope": "https://dashscope.aliyuncs.com/compatible-mode/v1",
    "doubao": "https://ark.cn-beijing.volces.com/api/v3",
    "kimi": "https://api.moonshot.cn/v1",
    "minimax": "https://api.minimax.chat/v1",
}

KNOWN_PROVIDERS = {"anthropic", "openai-compatible", "ollama", "mock", *CHINESE_PROVIDER_PRESET_BASE_URLS}


def _require_env(name: str) -> str:
    value = os.environ.get(name)
    if not value:
        raise RuntimeError(f"{name} environment variable is required for the selected MODEL_PROVIDER")
    return value


async def _real_ollama_post(url: str, json_body: dict[str, object]) -> httpx.Response:
    async with httpx.AsyncClient(timeout=60.0) as client:
        return await client.post(url, json=json_body)


def create_provider_from_env() -> TextCompletionProvider:
    provider_name = _require_env("MODEL_PROVIDER")
    if provider_name not in KNOWN_PROVIDERS:
        raise RuntimeError(
            f'Unknown MODEL_PROVIDER "{provider_name}" — expected one of: {", ".join(sorted(KNOWN_PROVIDERS))}'
        )

    if provider_name == "mock":
        return MockProvider()

    model = _require_env("MODEL_NAME")

    if provider_name == "anthropic":
        api_key = _require_env("MODEL_API_KEY")
        client = AsyncAnthropic(api_key=api_key, base_url=os.environ.get("MODEL_BASE_URL"))
        return AnthropicProvider(client, model=model)

    if provider_name == "ollama":
        base_url = _require_env("MODEL_BASE_URL")
        return OllamaProvider(_real_ollama_post, base_url=base_url, model=model)

    api_key = _require_env("MODEL_API_KEY")
    base_url = os.environ.get("MODEL_BASE_URL") or CHINESE_PROVIDER_PRESET_BASE_URLS.get(provider_name)
    client = AsyncOpenAI(api_key=api_key, base_url=base_url)
    return OpenAICompatibleProvider(client, model=model)
